src/main/java/com/salesforce/dynamodbv2/mt/util/CachingAmazonDynamoDbStreams.java

```diff
@@ -801,66 +664,25 @@ private GetRecordsResult getRecords(CachingShardIterator iterator) {
             }
 
-            List<Record> loadedRecords = loadedRecordsResult.getRecords();
-            String loadedNextIterator = loadedRecordsResult.getNextShardIterator();
-
-            // if we didn't load anything, return without adding cache segment (preserves non-empty range invariant)
-            // TODO could cache empty results for (short) time period to avoid having every client hit the stream.
-            if (loadedRecords.isEmpty()) {
-                // end of shard, just return
-                if (loadedNextIterator == null) {
-                    return loadedRecordsResult;
-                }
-                // otherwise compute next iterator (update cache for lazy iterators)
-                CachingShardIterator nextIterator;
-                if (iterator.getDynamoDbIterator().isPresent()) {
+            // compute next iterator
+            final String loadedNextIterator = result.getNextShardIterator();
+            final List<Record> records = result.getRecords();
+            final CachingShardIterator nextIterator;
+            if (loadedNextIterator == null) {
+                nextIterator = null;
+            } else {
+                if (records.isEmpty() && iterator.getDynamoDbIterator().isPresent()) {
                     nextIterator = iterator.withDynamoDbIterator(loadedNextIterator);
                 } else {
-                    iteratorCache.put(iterator, loadedNextIterator);
-                    nextIterator = iterator;
+                    nextIterator = iterator.nextShardIterator(records);
+                    iteratorCache.put(nextIterator, loadedNextIterator);
                 }
-                return new GetRecordsResult()
-                    .withRecords(loadedRecords)
-                    .withNextShardIterator(nextIterator.toExternalString());
-            }
-
-            // update iterator cache
-            if (loadedNextIterator != null) {
-                iteratorCache.put(iterator.nextShardIterator(loadedRecords), loadedNextIterator);
             }
 
-            // update record cache
-            // First resolve iterator position (either to sequence number it specifies or first record sequence number)
-            IteratorPosition loadedPosition = iterator.resolvePosition(loadedRecords);
-            Optional<GetRecordsResult> cachedResult;
-            GetRecordsResult result;
-            final Lock writeLock = recordsCacheLock.writeLock();
-            writeLock.lock();
-            try {
-                // Add retrieved records to cache under resolved position
-                cachedResult = addToCache(loadedPosition, loadedRecordsResult);
-
-                // now lookup result: may not be exactly what we loaded if we merged result with other segments.
-                // can also be empty if the cache is too small to hold the result, in which case we just return loaded.
-                result = getFromCache(loadedPosition).orElseGet(() ->
```

Comment:
This behavior has changed as well: instead of adding to the cache and then doing another cache lookup, the new impl adds the retrieved records the records it found in the cache (without an extra lookup).